### PR TITLE
net: lwm2m: add notification storing/retrieving API

### PR DESCRIPTION
**Pull request by marinjurjevic**
# Background

For last couple of months there's been a lot of discussion how to approach the general problem of storing notifications when, for whatever reason, they can not be delivered to server. There has been sporadic  discussion on Discord, though it's been quite some time. Original issue is still open here on #32610.

A couple of PR's have been merged which add some reporting infrastructure which was missing (reporting notification timeout/ACK and observer added/removed events). Look https://github.com/zephyrproject-rtos/zephyr/pull/36453 and https://github.com/zephyrproject-rtos/zephyr/pull/39651. This PR does not include the latter one as it conflicts with implementation and I haven't had time to see if it can be rebased onto it. 

# Approach
Approach implemented in this PR has been mentioned on Discord's lwm2m channel, but it mostly revolves around idea presented in my comment in #32160 [here](https://github.com/zephyrproject-rtos/zephyr/issues/32610#issuecomment-942238923).

It is counterpart to another approach, in which most of the processing is done by application logic instead of the engine. 

TLDR;
Idea is to have automatic notification storing whenever notification fails to be delivevered to server. Retrieving is triggered from application space using trigger function/event which would leave some control in hands of application regarding when will the engine start popping and sending stored notifications. Main advantage is that application only has to implement API (e.g. using flash circular buffer or RAM buffer as presented in the sample - which is naive array for simplicity, but ring buffer could also be used).

# Implementation
Implementation is simple and and naive as this is just a draft. API is defined in FIFO manner for storing, popping and checking if buffer is empty. When notification timeout callback is executed, notification structure is filled with relevant information (path information, encoded payload data and length) and pushed to storage. Retrieving is done by utilizing system workqueue which will pop elements until there is no more notifications or there is no more space in in-flight buffers. In case generating message fails, notification is pushed back to storage and work is delayed for 1 second. During testing, the only reason this can fail is when all in-flight buffers are exhausted, but this needs more robust handling.

# Pros
- Simplicity in application by moving all the logic to lwm2m engine which would otherwise clutter application
- Changing storage medium is just matter of providing different implementation (also we could have flash as default selected as it will probably be the most used one)
- No need for keeping track of notifications, observers, ACKs, timeouts etc 
 
# Cons
- FIFO structure - requires all observers be registered before retrieval phase (could be improved with different API contract)
- Less flexibility in the application code

# Improvements
- [x] Currently notifs are only stored after timeout. Specs also indicate disable and offline state should handle it, but these are not implemented in the engine yet)
  Implemented client connection status {CONNECTED, DISABLED and OFFLINE}. This was necessary to avoid removing observers when registration update fails. 
- [ ] Handling in-flight buffer exhaustion
  This might not be a critical problem in initial draft. During notification retrieval phase, if lwm2m message can not be allocated, retrieval is postponed for 1 second. This results in pattern where notifications exhaust all in-flight buffers fast - you can see `CONFIG_LWM2M_ENGINE_MAX_PENDING` packets sent quickly and 1 second delay kicks in. 1 second is arbitrary, we should probably further reduce it to milliseconds range.
  If registration/update/bootstrap fails to allocate because of exhaustion, network error is reported and engine restarts registration. I think best solution would be to split internal buffers for Registration Interface and Information Reporting Interface so there is no clash between them. Also, for Registration Interface, one buffer would probably be enough as they never happen concurrently.
- [ ] Dropped in-flight notifications (add peek and pop only when notif gets ACK'd)
  API retrieval function is designed by pop method -> this implies when retrieved, notification is also deleted from storage. This means there is a window where it can get lost (e.g. it's sent but ACK never comes). If device reboots unexpectedly, data is lost forever. 
  NOTE: lwm2m sample uses FCB implementation, which actually doesn't delete notification, which means on reboot all notifications from previous run will be available. But this is not expected behavior per API pop contract.
 Another case this can happen is for example, reg. update fails before notification timeout. In this case, new registration wipes all in-flight buffers.
- [ ] Modify API from FIFO to dictionary - this would allow checking and retrieving notifs directly after observer is added
- [ ] Make this approach fully optional (using CONFIG_LWM2M_NOTIF_STORING_SUPPORT)
- [x] Add default storage mediums (FCB and RAM rinbufers)
  Sample uses FCB implementation.
-  [ ] Refactor generate_notify_message (decouple data retrieval in `do_read_op`)
  Compliance Checks should also pass after this is done
- [x] Refactor `lwm2m_obj_path` -> make it public (already done in https://github.com/zephyrproject-rtos/zephyr/pull/39651)

---

The optimal solution would be if this approach is possble to live side by side with it's counterpart and can be selectable through Kconfig.

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 1]
```diff
@@ -4151,6 +4168,20 @@ static int32_t retransmit_request(struct lwm2m_ctx *client_ctx,
 	return next_retransmission;
 }
 
+static int lwm2m_store_notification(struct lwm2m_message *msg)
+{
+	struct lwm2m_notif notif;
+
+	// header length + options length + payload marker (+1 byte)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4175]: C style comments (`/* */`) are required by Zephyr.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -4161,6 +4192,16 @@ static void notify_message_timeout_cb(struct lwm2m_message *msg)
 		}
 	}
 
+	int err = lwm2m_store_notification(msg);
+	if(err)
+	{
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4197]:
```suggestion
	if (err) {
```

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 3]
```diff
@@ -4161,6 +4192,16 @@ static void notify_message_timeout_cb(struct lwm2m_message *msg)
...
+		LOG_ERR("Error storing notification: %d", err);
+	}
+	else
+	{
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4201]:
```suggestion
	} else {
```

File: subsys/net/lib/lwm2m/Kconfig [hunk 4]
```diff
@@ -17,6 +17,14 @@ module-dep = LOG
 module-str = Log level for LWM2M library
 source "subsys/net/Kconfig.template.log_config.net"
 
+config LWM2M_NOTIF_STORING_SUPPORT
+	bool "Enable notification storing when Disabled or Offline"
+	default n
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:22]: No need for `default n` (it's implicit).

File: subsys/net/lib/lwm2m/Kconfig [hunk 5]
```diff
@@ -17,6 +17,14 @@ module-dep = LOG
...
+	help
+		If true, LWM2M engine will use general API for storing and
+		retrieving notifications from non-volatile storage. It's up to
+		user to provide API implementation
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/Kconfig:26]:
```suggestion
	  If true, LWM2M engine will use general API for storing and
	  retrieving notifications from non-volatile storage. It's up to
	  user to provide API implementation
```

File: samples/net/lwm2m_client/src/lwm2m-client.c [hunk 6]
```diff
@@ -454,6 +456,76 @@ static void rd_client_event(struct lwm2m_ctx *client,
...
+	counter++;
+
+
+	uint32_t random_int = sys_rand32_get();
+	double temp = random_int % 100;
+	LOG_WRN("Setting new temperature: %d", random_int%100);
+	lwm2m_engine_set_float("/3303/0/5700", &temp);
+}
+
+K_TIMER_DEFINE(my_timer, my_timer_handler, NULL);
```
Comment (rlubos) [inline samples/net/lwm2m_client/src/lwm2m-client.c:475]: I think we should use more meaningful names than `my_timer`, `my_timer_handler`.

**Review by rlubos — COMMENTED** [review]
Could you rebase your PR for starters? I think it's in conflict with changes from https://github.com/zephyrproject-rtos/zephyr/pull/39651 which was merged recently.

**Comment by rlubos** [comment]
> Could you rebase your PR for starters? I think it's in conflict with changes from #39651 which was merged recently.

> This PR does not include the latter one as it conflicts with implementation and I haven't had time to see if it can be rebased onto it.

Oh, I now see you are already aware of that.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 7]
```diff
@@ -1026,6 +1029,18 @@ static int lwm2m_send_message(struct lwm2m_message *msg)
 	return 0;
 }
 
+int lwm2m_dump_notifications(struct lwm2m_ctx *client_ctx)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:1032]: I think `dump` might be a bit confusing in this case, as it might suggest we're dropping the stotred notifications, not sending them. Perhaps `resend` would be more appropriate.

**Review by rlubos — COMMENTED** [review]
I see the downside of such an approach as it does not allow to combine multiple notifications into a single one, but I guess it's acceptable if we agree upon it.

**Comment by rlubos** [comment]
FYI @MaikVermeulen @JPSELC

**Comment by rlubos** [comment]
Also, another thing to consider is that an outdated notification is expected to have a timestamp field in the JSON payload, I don't see it covered in this draft proposal.

**Comment by JPSELC** [comment]
Thanks @marinjurjevic for taking this up and @rlubos for notifying me.
Apologies, I have not been able to work on this for some time now, but I expect to be returning to it next year.
This approach could work for me as my application's payload is an opaque resource which includes a timestamp. However, I think the amount of communication would be excessive (send and ack of each notification) if offline for a while, so ideally would prefer if notifications were combined, which would result in a single notification message for all historical notifications for that observer.
I've no objection to this approach though as an option, and appreciate all the effort that has gone into developing it.

**Comment by marinjurjevic** [comment]
> at an outdated notification is expected to have a timestamp field in the JSON payload, I don't see it covered in this draft proposal

Yes, I've only tested with TLV. Another downside is that notifications will be stored encoded with current content format. If resource/object is reobserved using other content format, that will not be respected (but this doesn't seem to be big deal).

**Comment by marinjurjevic** [comment]
> > Could you rebase your PR for starters? I think it's in conflict with changes from #39651 which was merged recently.
> 
> > This PR does not include the latter one as it conflicts with implementation and I haven't had time to see if it can be rebased onto it.
> 
> Oh, I now see you are already aware of that.

Sorry if I wasn't totally clear, this is still WIP. I didn't have time to clean up the code (this includes documentation - doxygen, refactoring symbol names (structures, functions etc.). I will still be doing some modifications, so there's no need to do any extensive review at this phase. After New Year, I will work on this extensively, and then I'll have more information. I just wanted to get this into public so it gets some views and general feeling if idea is sound and will be considered for merging. All feedback is appreciated, and I would like to get as much of it I can before proposing the final idea/implementation.

**Comment by marinjurjevic** [comment]
Here is the summary of latest changes:
- Rebased upon https://github.com/zephyrproject-rtos/zephyr/pull/39651
- Cleaned up code (formatting, logs)
- Use existing coap API functions for parsing and appending payload
- Removed temperature value simulation, since it's not necessary (Device timestamp resource is already updated every 10 seconds)

**Comment by marinjurjevic** [comment]
I updated original post with checkboxes regarding completion and short decriptions/use cases. 

Latest updates include:
- Client connection status (we must not delete observers when restarting registration)
- Add FCB based implementation of storing API
- Minor bugfixes

---

This PR is getting rather heavy at this point, maybe it's good idea to split it up? 
I've been squashing all commits into one, should this be done like this - do PR have to contain only one commit?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 8]
```diff
@@ -4435,8 +4590,12 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 
 void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx)
 {
-	sys_slist_init(&client_ctx->pending_sends);
-	sys_slist_init(&client_ctx->observer);
+	if (client_ctx->conn_status == LWM2M_CLIENT_CONNECTED) {
```
Comment (marinjurjevic) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4593]: This block ensures observer list remains intact when this function is eventually called from `lwm2m_engine_start` upon registration/update failure. It also preserves reinitializing pending list. Both of these lists are also NOT cleared in `lwm2m_engine_context_close` function as part of same goal. Observers must be kept intact so notifications can be generated normally. 
The reason for keeping `pending_sends` intact as well is mentioned in the original post:
> Another case this can happen is for example, reg. update fails before notification timeout. In this case, new registration wipes all in-flight buffers.
To prevent this, we leave the `pending_sends` as is so the notifications which are in-flight continue to be transmitted and timeout eventually. After they timeout, they will be pushed to storage medium. There is one minor problem with this approach and that is order in which these notifications will be stored.

**Comment by marinjurjevic** [comment]
> @marinjurjevic can you rebase this so we get a chance at reviewing and merging this?

Hi, I'm not sure if this PR is still relevant. This PR content was used as series of patches on a project which I'm not a part of anymore. 
IIRC, this PR would introduce storing serialized notifications in arbitrary storage. There was never a consensus that this is best approach, some people preferred to not store serialized notifications .
Also this PR includes some minor changes to LwM2M engine itself, which I think have been solved in the meantime.

@rlubos Are you still maintaining LwM2M? Do you have some information has there been any progress regarding notification storing?

**Comment by rlubos** [comment]
Hello @marinjurjevic, thanks for the prompt reply!

> @rlubos Are you still maintaining LwM2M? Do you have some information has there been any progress regarding notification storing?

Technically yes, but now with great help from @SeppoTakalo and his team.

Regarding notifications with historical data, Zephyr LwM2M implementation took a different approach than proposed in this PR. Instead of storing serialized notifications, a [resource data cache](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/Kconfig#L212) was introduced which allows to store timestamped historical values of resources. This allows to send historical data when the device is back online, when notification is generated. Therefore, it seems there's no longer a need to store notifications themselves.

I suggest hereby that we close the PR, since there is an alternative for notification storing.
